(DOCS) Update CLI reference and conceptual docs for v3.0.0 (#672)
PowerShell/DSC: docs/concepts/configuration-documents/overview.md

---
description: >-
  DSC configuration documents are YAML or JSON data files that define the desired state of a system
  declaratively. They're used to retrieve, validate, and enforce the state of multiple resource
  instances.
ms.date: 03/25/2025
title: DSC configuration documents
---

# DSC configuration documents

<!-- markdownlint-disable MD013 -->

In Microsoft's Desired State Configuration (DSC) platform, DSC configuration documents declare the
desired state of a system as data files. Configuration documents define a collection of
[DSC resource][01] instances to describe what the desired state should be, not how to put the
system into that state. The DSC resources handle the _how_ for each instance.

DSC can process configuration documents to:

- Retrieve the current state of the defined resource instances with the `dsc config get` command.
- Validate whether the instances are in the desired state with the `dsc config test` command.
- Enforce the desired state for the instances with the `dsc config set` command.
- Export a new configuration document with every instance of a set of resources with the
  `dsc config export` command.

> Configuration documents are YAML or JSON files that contain a single object. The object's

Reviewer: Does it help to mention relationship to ARM templates since those have more docs and examples?

Reviewer 2: And given the very positive reaction to Bicep at PS Summit, this makes it sound like we _only_ support the two languages, when we actually support a third, as an abstraction of the JSON.

> Configuration documents are YAML or JSON files that contain a single object. The object's

Reviewer 2: ```suggestion
Configuration documents can be formatted as YAML, JSON, or Bicep (which compiles to JSON), and contain a single object. The object's
```

  resource's instance schema.

Configuration documents can't include the same instance more than once. Declaring the same instance
with different properties leads to enforcement cycling, where each declaration enforces an
incompatible state for the instance on every run.

## Getting the current state of a configuration

The `dsc config get` command retrieves the current state of the resource instances defined in a
configuration document. When you use this command, DSC also:

- Collects any message emitted by the resources during the operation.
- Indicates whether any of the resources raised an error.
- Provides metadata about the operation as a whole and for each resource instance.

```sh
dsc config get --file ./example.config.dsc.yaml
```

```yaml
metadata:
  Microsoft.DSC:
    version: 3.0.0
    operation: get
    executionType: actual
    startDatetime: 2025-02-24T16:09:40.671454400-06:00
    endDatetime: 2025-02-24T16:09:41.850086300-06:00
    duration: PT1.1786319S
    securityContext: restricted
results:
- metadata:
    Microsoft.DSC:
      duration: PT0.2751153S
  name: example key value
  type: Microsoft.Windows/Registry
  result:
    actualState:

>       keyPath: HKCU\example\key

Reviewer: Enclose in single quotes so that the backslash doesn't need to be escaped

```suggestion
      keyPath: 'HKCU\example\key'
```

Author: The backslash doesn't need escaping in YAML for unquoted strings:

```console
@'
unquoted:       foo\bar\baz
double_quoted: "foo\\bar\\baz"
single_quoted: 'foo\bar\baz'
'@ | ConvertFrom-Yaml | Format-List

Name  : unquoted
Value : foo\bar\baz

Name  : double_quoted
Value : foo\bar\baz

Name  : single_quoted
Value : foo\bar\baz
```

Reviewer: I guess it's just GitHub's syntax coloring then

      _exist: false
messages: []
hadErrors: false
```

## Testing a configuration

The `dsc config test` command compares the current state of the resource instances to their desired
state. The result for each instance includes:

- The desired state for the instance.
- The actual state of the instance.
- Whether the instance is in the desired state.
- The list of properties that are out of the desired state, if any.

When you use this command, DSC also:

- Collects any message emitted by the resources during the operation.
- Indicates whether any of the resources raised an error.
- Provides metadata about the operation as a whole and for each resource instance.

```sh

> dsc config test --file /example.config.dsc.yaml

Reviewer: ```suggestion
dsc config test --file ./example.config.dsc.yaml
```

Is this missing the period?  Otherwise it's an absolute path

Reviewer: I don't see this as fixed?
